hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTaskWithLegacy.java

```diff
@@ -0,0 +1,278 @@
+    Map<Long, NSSummary> nsSummaryMap = new HashMap<>();
+
+    try {
+      // reinit Recon RocksDB's namespace CF.
+      getReconNamespaceSummaryManager().clearNSSummaryTable();
+
+      Table keyTable = omMetadataManager.getKeyTable(bucketLayout);
+
+      TableIterator<String, ? extends Table.KeyValue<String, OmKeyInfo>>
+          keyTableIter = keyTable.iterator();
+
+      while (keyTableIter.hasNext()) {
+        Table.KeyValue<String, OmKeyInfo> kv = keyTableIter.next();
+        OmKeyInfo keyInfo = kv.getValue();
+
+        if (keyInfo != null) {
+          setKeyParentID(keyInfo);
+
+          if (keyInfo.getKeyName().endsWith(OM_KEY_PREFIX)) {
+            OmDirectoryInfo directoryInfo =
+                new OmDirectoryInfo.Builder()
+                    .setName(keyInfo.getKeyName())
+                    .setObjectID(keyInfo.getObjectID())
+                    .setParentObjectID(keyInfo.getParentObjectID())
+                    .build();
+            handlePutDirEvent(directoryInfo, nsSummaryMap);
+          } else {
+            handlePutKeyEvent(keyInfo, nsSummaryMap);
+          }
+        } else {
+          LOG.error("Reprocess KeyInfo for NSSummaryTaskWithLegacy is null");
+        }
+      }
+    } catch (IOException ioEx) {
+      LOG.error("Unable to reprocess Namespace Summary data in Recon DB. ",
+          ioEx);
+      return new ImmutablePair<>(getTaskName(), false);
+    }
+
+    try {
+      writeNSSummariesToDB(nsSummaryMap);
+    } catch (IOException e) {
+      LOG.error("Unable to write Namespace Summary data in Recon DB.", e);
+      return new ImmutablePair<>(getTaskName(), false);
+    }
+    LOG.info("Completed a reprocess run of NSSummaryTaskWithLegacy");
+    return new ImmutablePair<>(getTaskName(), true);
+  }
+
+  private void setKeyParentID(OmKeyInfo keyInfo) throws IOException {
+    String[] keyPath = keyInfo.getKeyName().split(OM_KEY_PREFIX);
+
+    //if (keyPath > 1) there is one or more directories
+    if (keyPath.length > 1) {
+      StringBuilder bld = new StringBuilder();
+      for (int i = 0; i < keyPath.length - 1; i++) {
+        bld.append(keyPath[i])
+            .append(OM_KEY_PREFIX);
+      }
+      String parentKeyName = bld.toString();
+      String keyBytes =
+          reconOMMetadataManager.getOzoneKey(keyInfo.getVolumeName(),
+              keyInfo.getBucketName(), parentKeyName);
+      OmKeyInfo parentKeyInfo = reconOMMetadataManager
+          .getKeyTable(bucketLayout)
+          .get(keyBytes);
+
+      if (parentKeyInfo != null) {
+        keyInfo.setParentObjectID(parentKeyInfo.getObjectID());
+      } else {
+        LOG.error("ParentKeyInfo for NSSummaryTaskWithLegacy is null");
+      }
+    } else {
+      String bucketKey = reconOMMetadataManager
+          .getBucketKey(keyInfo.getVolumeName(), keyInfo.getBucketName());
+      OmBucketInfo parentBucketInfo =
+          reconOMMetadataManager.getBucketTable().get(bucketKey);
+
+      if (parentBucketInfo != null) {
```

Comment:
Same, It's a bad error if this is null, and we should throw the exception, so remove the check please.